ksp_plugin_adapter/burn_editor.cs

```diff
@@ -28,51 +31,195 @@ class BurnEditor {
     UnityEngine.GUILayout.EndVertical();
     UnityEngine.GUI.skin = old_skin;
-    return changed;
+    return changed && enabled;
   }
 
-  // Returns the equivalent of the .NET >= 4 format
-  // span.ToString(@"ddd \d hh \h mm \m\i\n ss.FFF \s").
-  private string FormatTimeSpan (TimeSpan span) {
-     return span.Days.ToString("000") + " d " +
-            span.Hours.ToString("00") + " h " +
-            span.Minutes.ToString("00") + " min " +
-            span.Seconds.ToString("00") + " s";
+  public void Reset(NavigationManoeuvre manoeuvre) {
+    Burn burn = manoeuvre.burn;
+    Δv_tangent_.value = burn.delta_v.x;
+    Δv_normal_.value = burn.delta_v.y;
+    Δv_binormal_.value = burn.delta_v.z;
+    initial_time_.value = burn.initial_time;
+    reference_frame_selector_.Reset(burn.frame);
+    duration_ = manoeuvre.duration;
+  }
+
+  public Burn Burn() {
+    return new Burn{
+        thrust_in_kilonewtons = thrust_in_kilonewtons_,
+        specific_impulse_in_seconds_g0 = specific_impulse_in_seconds_g0_,
+        frame = reference_frame_selector_.FrameParameters(),
+        initial_time = initial_time_.value,
+        delta_v = new XYZ{x = Δv_tangent_.value,
+                          y = Δv_normal_.value,
+                          z = Δv_binormal_.value } };
+  }
+
+  public void ReferenceFrameChanged(NavigationFrameParameters parameters) {
+    changed_reference_frame_ = true;
+  }
+
+  public void Close() {
+    reference_frame_selector_.Dispose();
+  }
+
+  private void ComputeEngineCharacteristics() {
+    ModuleEngines[] active_engines =
+        (from part in vessel_.parts
+         select (from PartModule module in part.Modules
+                 where module is ModuleEngines &&
+                       (module as ModuleEngines).EngineIgnited
+                 select module as ModuleEngines)).SelectMany(x => x).ToArray();
+    Vector3d reference_direction = vessel_.ReferenceTransform.up;
+    double[] thrusts =
+        (from engine in active_engines
+         select engine.maxThrust *
+             (from transform in engine.thrustTransforms
+              select Math.Max(0,
+                              Vector3d.Dot(reference_direction,
+                                           -transform.forward))).Average()).
+            ToArray();
+    thrust_in_kilonewtons_ = thrusts.Sum();
+
+    // This would use zip if we had 4.0 or later.  We loop for now.
+    double Σ_f_over_i_sp = 0;
+    for (int i = 0; i < active_engines.Count(); ++i) {
+      Σ_f_over_i_sp +=
+          thrusts[i] / active_engines[i].atmosphereCurve.Evaluate(0);
+    }
+    specific_impulse_in_seconds_g0_ = thrust_in_kilonewtons_ / Σ_f_over_i_sp;
+
+    // If there are no engines, fall back onto RCS.
+    if (thrust_in_kilonewtons_ == 0) {
+      engine_warning_ = "No active engines, falling back to RCS";
+      ComputeRCSCharacteristics();
+    }
+  }
+
+  private void ComputeRCSCharacteristics() {
+    ModuleRCS[] active_rcs =
```

Comment:
Is there a way that these linq queries could be factored out with the ones in the previous functions?